### ISSUE TITLE
Use the config directory indicated on pillar and not the real_config value

### ISSUE THREAD

**Issue by richerve** [saltstack-formulas/hadoop-formula#44]
Hi, Is there any specific reason why the formula uses "alt_home" and "real_home" and the corresponding "_config" ?
I can understand that maybe you want to have several versions and manage them with the alternatives system but for me seems counter intuitive to specify the parameter in pillar and see other directory created. I'll give you an example:

**Pillar**

``` yaml
hadoop:
  version: apache-2.7.3
  prefix: /usr/lib/hadoop/hadoop-2.7.3
  config:
    directory: /etc/hadoop
```

What I'm trying to do with this is having /usr/lib/hadoop/hadoop-2.7.3 as a prefix and in a upgrade apply the same formula with hadoop-2.8.0 for example and have the directory inside /usr/lib/hadoop. Then the alternatives system links the binaries to /usr/bin as it does now.

On the configuration side I only want to have /etc/hadoop without any versioning name (similar to what the formula is doing now with "/etc/hadoop-2.7.3" but without the version). I want it this way because on the execution side the alternatives system only handles "one at a time" so is the administrator responsibility to maintain a config compatible with that binary.

If you don't agree with my way of seeing the configuration then either way my main point is that the formula has side-effects and the information on the pillar doesn't represents exactly the actions of it.

I can make the changes to use the information provided by the pillar and don't use the "alt" "real" options if it's an agreed decision at some point.

**Comment by sroegner** [comment]
The reason for the way it is is mostly that it came together the other way round - the alt_and real_ directories were somewhat reflecting things that Ambari and Cloudera Manager did at the time and the alternatives links into /usr/bin came much later.

Having versioned conf is IMO desirable to be able to go back and forth between versions _and_ doing forensics after doing so if necessary. Far-fetched? Sure. This could be made simpler but the chances of breaking existing installations are pretty high - we rely heavily on hadoop-formula for everyday test environments and see problems with almost every other change.

Just my two cents.

**Comment by richerve** [comment]
I fully understand you and I saw that "pattern" in other java developers and sysadmins. I also agree if you are not using a CM system because is easier in some sense to change a symlink or replace a directory with another. I also know that this would represent a total change in the interface of the formula and I started the issue just to maintain at least the discussion open. Meanwhile I am refactoring the formula in a separate branch on my fork.

Just to leave my two cents on this topic, I think that using saltstack means to maintain the versioning in the interface closer to the user that in this case is the pillar, this means maintaining a pillar with the configuration changes for each version and be able to apply one or the other meanwhile the formula does the same thing without side-efects. That is why almost all CM's are created with the idea of states and are idempotent.

Thank you for replying

**Comment by sroegner** [comment]
@richerve I fully agree to what you said. I believe that my deviating from this idea of where the conf data should live came from my lack of experience at the time of the original implementation as well as from pillars not quite being where they are today - the hadoop formula was started right after formulas had become a thing. Also consider an organization where there is one or two people with full understanding of the cm system and ability to debug problems being part of a much larger group of admins trying to make sense of just the result - in an observer role if you will - who might have no visibility into cm code. There is another driver for this type of implementation.

I am still looking forward to seeing the result of your refactoring effort ;-)

**Comment by blbradley** [comment]
This is a great issue! zookeeper-formula does something similar, and I was about to file a similar issue there.

@sroegner I was a bit unclear on what behavior you are interested in. Are you in favor of dropping `alt_*` and `real_*`? Or do you still need that usage?

**Comment by sroegner** [comment]
@blbradley I would say i am in favor of keeping things simple. The similarity between zookeeper and hadoop formula simply come from the fact that i originally wrote both. The question IMO has to be whether the gain in simplicity justifies the potential migration pain in existing installations.

What _actual_ issue beside it being unnecessary are you seeing?

**Comment by blbradley** [comment]
Yes! The lack of documentation. There is nothing in this formula nor zookeeper-formula that describes the intend of those. I suppose it was confusing enough for me two years ago that I extended all the related states in zookeeper-formula to use `real_home` and `real_config`.

P.S. I don't have a strong opinion on this. Just seeking to understand the 'why'.

**Comment by sroegner** [comment]
Sorry to hear it was/is that confusing. Once we have travis automation in place we can work on it.

**Comment by blbradley** [comment]
Two years ago, my technique was probably due to lack of experience with SaltStack or trying to match what Debian's packaging does. But, recently, I did look through the histories of zookeeper-formula and this formula in order to determine the 'why' of these parameters. I traced them back to the initial commits! :)

These parameters could be kept but need the 'how' in the docs. At the least, a pillar example.

Thanks @sroegner!
